contracts/src/Nodes.sol

```diff
@@ -1,91 +1,162 @@
 // SPDX-License-Identifier: MIT
 pragma solidity ^0.8.13;
 
+import "@openzeppelin/contracts/token/ERC721/ERC721.sol";
 import "@openzeppelin/contracts/access/Ownable.sol";
 
-contract Nodes is Ownable {
-    constructor() Ownable(msg.sender) {}
+/**
+A NFT contract for XMTP Node Operators.
 
+The deployer of this contract is responsible for minting NFTs and assinging them to node operators.
+
+All nodes on the network periodically check this contract to determine which nodes they should connect to.
+ */
+contract Nodes is ERC721, Ownable {
+    constructor() ERC721("XMTP Node Operator", "XMTP") Ownable(msg.sender) {}
+
+    // uint16 counter so that we cannot create more than 65k IDs
+    // The ERC721 standard expects the tokenID to be uint256 for standard methods unfortunately
+    uint16 private _nodeIdCounter;
+
+    // A node, as stored in the internal mapping
     struct Node {
+        bytes signingKeyPub;
         string httpAddress;
-        uint256 originatorId;
         bool isHealthy;
-        // Maybe we want a TLS cert separate from the public key for MTLS authenticated connections?
     }
 
-    event NodeUpdate(
-        bytes publicKey,
-        string httpAddress,
-        uint256 originatorId,
-        bool isHealthy
-    );
+    struct NodeWithId {
+        uint16 nodeId;
+        Node node;
+    }
 
-    // List of public keys
-    bytes[] publicKeys;
+    event NodeUpdated(uint256 nodeId, Node node);
 
-    // Mapping of publicKey to node
-    mapping(bytes => Node) public nodes;
+    // Mapping of token ID to Node
+    mapping(uint256 => Node) private _nodes;
 
     /**
-    Add a node to the network
+    Mint a new node NFT and store the metadata in the smart contract
      */
     function addNode(
-        bytes calldata publicKey,
+        address to,
+        bytes calldata signingKeyPub,
+        string calldata httpAddress
+    ) public onlyOwner returns (uint16) {
+        uint16 nodeId = _nodeIdCounter;
+        _mint(to, nodeId);
+        _nodes[nodeId] = Node(signingKeyPub, httpAddress, true);
+        _emitNodeUpdate(nodeId);
+        _nodeIdCounter++;
+
+        return nodeId;
+    }
+
+    /**
+    Override the built in transferFrom function to block NFT owners from transferring
+    node ownership.
+
+    NFT owners are only allowed to update their HTTP address and MTLS cert.
+     */
+    function transferFrom(
+        address from,
+        address to,
+        uint256 tokenId
+    ) public override {
+        require(
+            _msgSender() == owner(),
+            "Only the contract owner can transfer Node ownership"
+        );
```

Comment:
Small nit, any reason this one uses a require statement, but `updateHealth` uses `public onlyOwner`?

Reply:
Only that I wanted a better error message, but agree the inconsistency feels weird